runtime/common/src/apis.rs

```diff
@@ -345,22 +345,16 @@ macro_rules! impl_runtime_apis_plus_common {
 
 					// Because the staking solution calculates the next staking set at the beginning
 					// of the first block in the new round, the only way to accurately predict the
-					// authors would be to run the staking election while predicting. However this
-					// election is heavy and will take too long during prediction. So instead we
-					// work around it by always authoring the first slot in a new round. A longer-
-					// term solution will be to calculate the staking election result in the last
-					// block of the ending round.
-					if parachain_staking::Pallet::<Self>::round().should_update(block_number) {
-						log::info!(target: "nimbus-staking-workaround", "A new round is starting.\
-						Moonbeam will author during this slot without predicting eligibility first.\
-						You may see a `CannotBeAuthor` error soon. This is expected and harmless.\
-						It will be resolved soon.");
-
-						true
-					}
-					else {
+					// authors is to compute the staking selection while predicting.
+					let mut round = parachain_staking::Pallet::<Self>::round();
+					if round.should_update(block_number) {
+						// get top collator candidates for expected selection to occur
+						let top_candidates = parachain_staking::Pallet::<Self>::compute_top_candidates();
+						use nimbus_primitives::AccountLookup;
+						let author_account_id = pallet_author_mapping::Pallet::<Self>::lookup_account(&author).expect("expect to have registered author mapping");
+						top_candidates.contains(&author_account_id)
```

Comment:
This isn't enough. This just ensures that the author in question is in the staked set. We need to also ensure that they are selected by the pseudo-random selection algorithm. `AuthorInherent::can_author(&author, &slot)` handles all of that, but requires that the new set be stored in storage. I wish I had thought of that earlier. I'm starting to wonder if we can avoid going to storage at all.

Reply:
Could we copy paste the pseudo-random selection algorithm here? Or extract the pseudo-random selection algorithm from `can_author` into its own public function that returns a `Vec<AccountId>` of the selected each block, so it can be called here without calling can_author?

Reply:
We definitely could do that. My concern is that we're starting to have to repeat a lot of logic that already lives in the author inherent pallet. See https://github.com/PureStake/cumulus/blob/nimbus/pallets/author-inherent/src/lib.rs#L185-L199

This bit does the lookup from the author mapping (in a way that is generic over how the mapping is constructed and stored), then calls the chain of all the nimbus filters. That definitely could be repeated here, but  I'm starting to think it makes more sense for our logic here to be:

```rust
// Initialize frame system
// Initial randomness collective flip
if new_round_starting {
  // Initialize parachain staking (to run the election)
}
// Call the already-written author checking logic from pallet author inherent
```

That pseudo code relies on writing the election result to a memory-backed storage overlay that will never go to disk (because it isn't on-chain).

Reply:
Another issue with copying the code here is that if we ever change our author selection logic, we'll have to change it in two places to keep it synced. If we stick to initializing pallets and calling author inherent here, then any authoring changes will be reflected here by default.

Reply:
> My concern is that we're starting to have to repeat a lot of logic that already lives in the author inherent pallet.

Right, it's the same logic. That's why we could extract it into a public function to call it here

> Another issue with copying the code here is that if we ever change our author selection logic, we'll have to change it in two places to keep it synced.

..and then just change that function body to keep them both synced.

Is the concern code duplication or efficiency? I think both can be resolved. When it comes to efficiency, It'd run the same logic in the other branch via `can_author` so it wouldn't be less efficient to run either.

Reply:
<going to show with code>

Reply:
My concern is code style. No concern about efficiency. Thanks for writing the code.